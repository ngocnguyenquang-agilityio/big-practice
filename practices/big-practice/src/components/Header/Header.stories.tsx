// Types
import type { Meta, StoryObj } from '@storybook/react';

// Components
import Header from '.';

export default {
  title: 'Components/Header',
  component: Header,
  render: () => <Header toggleCart={() => {}} />,
} as Meta;

type Story = StoryObj;

export const Default: Story = {
  render: () => <Header toggleCart={() => {}} />,
};
