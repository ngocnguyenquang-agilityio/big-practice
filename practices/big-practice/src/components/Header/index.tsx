// Libs
import { ChangeEvent, useCallback, useState } from 'react';

// Icons
import { CartIcon, LogoIcon } from '@assets/icons';

// Components
import { Button } from '@components/Button';
import Cart from '@components/Cart';
import SearchInput from '@components/SearchInput';

// Store
import { useEndpoint } from '@stores/endpoint';

// TODO: Update type when implement function

const Header = () => {
  const [isOpenCart, setIsOpenCart] = useState(false);
  const { updateEndpoint } = useEndpoint();

  const handleSearch = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    const inputValue = e.target.value;
    setTimeout(() => {
      updateEndpoint(`/search?q=${inputValue}`);
    }, 1000);
  }, []);

  const handleToggleCart = () => {
    setIsOpenCart(!isOpenCart);
  };

  return (
    <header className='relative flex items-center justify-between p-4 lg:px-6'>
      <div className='flex w-full items-center'>
        <div className='flex w-full md:w-1/3 items-center'>
          <div className='mr-2 flex w-full items-center justify-center md:w-auto lg:mr-6'>
            <Button
              variant='link'
              size='icon'
            >
              <LogoIcon />
            </Button>
            <a
              className='ml-2 text-sm font-medium uppercase text-white'
              href=''
            >
              acme store
            </a>
          </div>
          <ul className='gap-6 text-sm md:flex md:items-center'>
            <li className='underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer'>
              All
            </li>
            <li className='underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer'>
              Shirts
            </li>
          </ul>
        </div>
        <div className='justify-center md:flex md:w-1/3'>
          <SearchInput
            type='text'
            placeholder='Search Products...'
            onChange={handleSearch}
          />
        </div>
        <div className='flex justify-end md:w-1/3'>
          {!isOpenCart ? (
            <Button
              variant='outline'
              size='icon'
              onClick={handleToggleCart}
            >
              <CartIcon />
            </Button>
          ) : (
            <Cart handleToggleCart={handleToggleCart} />
          )}
        </div>
      </div>
    </header>
  );
};

export default Header;
