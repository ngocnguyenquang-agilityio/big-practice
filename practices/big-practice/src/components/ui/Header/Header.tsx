// Components
import { CartIcon, LogoIcon } from "@components/icons"
import { Button } from "@components/ui/Button/Button"
import SearchInput from "@components/ui/SearchInput/SearchInput"

// TODO: Update type when implement function

const Header = () => {
  return (
    <div className="flex w-full items-center">
      <div className="flex w-full md:w-1/3 items-center">
        <div className="mr-2 flex w-full items-center justify-center md:w-auto lg:mr-6">
          <Button variant='link' size='icon'>
            <LogoIcon />
          </Button>
          <a className="ml-2 text-sm font-medium uppercase text-white" href="">acme store</a>
        </div>
        <ul className="gap-6 text-sm md:flex md:items-center">
          <li className="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer">
            All
          </li>
          <li className="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer">
            Shirts
          </li>
          <li className="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer">
            Stickers
          </li>
        </ul>
      </div>
      <div className="justify-center md:flex md:w-1/3">
        <SearchInput
          type="text"
          placeholder="Search Products..."
          onChange={() => { }}
        />
      </div>
      <div className="flex justify-end md:w-1/3">
        <Button variant='outline' size='icon'>
          <CartIcon />
        </Button>
      </div>
    </div>
  )
}

export default Header
