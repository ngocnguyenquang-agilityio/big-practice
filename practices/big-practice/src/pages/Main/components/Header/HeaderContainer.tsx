// Libs
import { ChangeEvent, memo, useCallback } from 'react';
import { Link, useNavigate, useSearchParams } from 'react-router-dom';

// Icons
import cartIcon from '@assets/cartIcon.svg';

// Components
import { Button } from '@components/Button/Button';
import { Icon } from '@components/Icon/Icon';
import SearchInput from '@components/SearchInput/SearchInput';

// Helpers
import { debounce } from '@helpers/utils';

// Constants
import { APP_ROUTERS, headerItems } from '@constants';

// Stores
import { useCartStore } from '@stores/cartStore';

// Types
import { IProduct } from '@interfaces';

export const Header = memo(
  ({
    cartItems,
    defaultValueSearch,
    numberOfItemsInCart,
    onToggleCart,
    handleSearch,
  }: {
    cartItems: IProduct[];
    defaultValueSearch: string;
    numberOfItemsInCart: number;
    onToggleCart: () => void;
    handleSearch: (e: ChangeEvent<HTMLInputElement>) => void;
  }) => {
    return (
      <header className='flex items-center p-4 lg:px-6'>
        <div className='flex w-full items-center justify-between '>
          <div className='flex sm:w-1/3 '>
            <Link
              to={APP_ROUTERS.HOMEPAGE}
              reloadDocument
              className='mr-2 flex w-full items-center justify-center sm:w-auto lg:mr-6'
            >
              <div className='flex flex-none items-center justify-center border border-neutral-700 bg-black h-[40px] w-[40px] rounded-xl'>
                <svg
                  name='logo-icon'
                  xmlns='http://www.w3.org/2000/svg'
                  aria-label='Acme Store logo'
                  viewBox='0 0 32 28'
                  className='h-5 w-5 fill-white h-[16px] w-[16px]'
                >
                  <path d='M21.5758 9.75769L16 0L0 28H11.6255L21.5758 9.75769Z' />
                  <path d='M26.2381 17.9167L20.7382 28H32L26.2381 17.9167Z' />
                </svg>
              </div>
              <span className='ml-2 text-sm font-medium uppercase text-white hidden lg:block'>acme store</span>
            </Link>
            <ul
              data-testid='header-items'
              className='hidden gap-6 text-sm lg:flex sm:items-center'
            >
              {headerItems.map((item) => (
                <Link
                  to={item.url}
                  key={item.url}
                  className='underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer'
                >
                  {item.label}
                </Link>
              ))}
            </ul>
          </div>

          <div className='justify-center sm:flex sm:w-1/3'>
            <form className='w-max-[550px] relative w-full lg:w-80 xl:w-full'>
              <SearchInput
                type='text'
                placeholder='Search Products...'
                onChange={handleSearch}
                defaultValue={defaultValueSearch}
              />
            </form>
          </div>

          <div className='flex justify-end sm:w-1/3'>
            <Button
              variant='outline'
              size='icon'
              onClick={onToggleCart}
            >
              <div className='relative flex h-11 w-11 items-center justify-center rounded-md border transition-colors border-neutral-700 text-white'>
                <Icon
                  svg={cartIcon}
                  name='cart-icon'
                />
                {cartItems.length > 0 && (
                  <div
                    data-testid='number-of-items'
                    className='absolute inline-flex items-center justify-center w-5 h-5 text-xs text-white bg-blue-600 rounded-full -top-2 -right-2 dark:border-gray-900'
                  >
                    {numberOfItemsInCart > 9 ? '9+' : numberOfItemsInCart}
                  </div>
                )}
              </div>
            </Button>
          </div>
        </div>
      </header>
    );
  },
);

const HeaderContainer = () => {
  const { cartItems, toggleCart } = useCartStore();

  const [searchParams] = useSearchParams();
  const navigate = useNavigate();

  const searchKeyword = searchParams.get('search');

  const handleSearch = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    const inputValue = e.target.value;

    navigate(`/?search=${inputValue.trim()}`);
  }, []);

  const searchWithDebounce = debounce(handleSearch, 1000);

  const numberOfItemsInCart = cartItems.reduce((total, currentItem) => (total += currentItem.quantity), 0);

  return (
    <Header
      cartItems={cartItems}
      defaultValueSearch={searchKeyword || ''}
      numberOfItemsInCart={numberOfItemsInCart}
      onToggleCart={toggleCart}
      handleSearch={searchWithDebounce}
    />
  );
};

export default HeaderContainer;
