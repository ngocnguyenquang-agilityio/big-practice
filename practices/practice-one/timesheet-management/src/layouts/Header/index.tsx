export const Header = (): React.ReactElement => (
  <header>
    <h1 className="text-4xl text-red-300 text text-center">Time Sheet</h1>
  </header>
);
