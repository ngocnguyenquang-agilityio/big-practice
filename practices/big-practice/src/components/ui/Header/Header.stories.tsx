import { Meta, StoryObj } from "@storybook/react";
import Header from "./Header";


export default {
  title: 'Components/Header',
  component: Header,
  render: () => <Header />
} as Meta

type Story = StoryObj

export const Default: Story = {
  render: () => <Header />
}
