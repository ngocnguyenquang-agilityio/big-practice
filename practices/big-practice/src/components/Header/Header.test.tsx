import { render } from '@testing-library/react';

// Components
import Header from '.';

// TODO: Update props and more test case when implement functions

describe('Header component', () => {
  test('Should render successfully', () => {
    const comp = render(<Header />)
    expect(comp).toBeTruthy()
  });

  test('Should match snapshot', () => {
    const comp = render(<Header />)
    expect(comp).toMatchSnapshot()
  });
});
