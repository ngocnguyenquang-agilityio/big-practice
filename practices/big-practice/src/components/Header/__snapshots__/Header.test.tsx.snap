// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex items-center justify-between p-4 lg:px-6"
      >
        <div
          class="flex w-full items-center"
        >
          <div
            class="flex w-full md:w-1/3 items-center"
          >
            <div
              class="mr-2 flex w-full items-center justify-center md:w-auto lg:mr-6"
            >
              <button
                class="inline-flex items-center text-white justify-center text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 underline-offset-4 hover:underline dark:text-neutral-100 h-11 w-11 rounded-md"
              >
                <div
                  class="flex flex-none items-center justify-center border border-neutral-700 bg-black h-[40px] w-[40px] rounded-xl"
                >
                  <svg
                    aria-label="Acme Store logo"
                    class="h-5 w-5 fill-white h-[16px] w-[16px]"
                    name="logo-icon"
                    viewBox="0 0 32 28"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M21.5758 9.75769L16 0L0 28H11.6255L21.5758 9.75769Z"
                    />
                    <path
                      d="M26.2381 17.9167L20.7382 28H32L26.2381 17.9167Z"
                    />
                  </svg>
                </div>
              </button>
              <a
                class="ml-2 text-sm font-medium uppercase text-white"
                href=""
              >
                acme store
              </a>
            </div>
            <ul
              class="gap-6 text-sm md:flex md:items-center"
            >
              <li
                class="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer"
              >
                All
              </li>
              <li
                class="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer"
              >
                Shirts
              </li>
            </ul>
          </div>
          <div
            class="justify-center md:flex md:w-1/3"
          >
            <div
              class="relative w-full"
            >
              <input
                class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 py-2 text-sm text-white placeholder:text-neutral-400"
                placeholder="Search Products..."
                type="text"
              />
              <div
                class="absolute inset-y-0 right-0 flex items-center pr-4 pointer-events-none"
              >
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 text-white"
                  fill="none"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="flex justify-end md:w-1/3"
          >
            <button
              class="inline-flex items-center text-white justify-center text-sm font-medium focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 border border-neutral-700 transition-colors h-11 w-11 rounded-md"
            >
              <svg
                class="lucide lucide-shopping-cart"
                fill="none"
                height="16"
                name="cart-icon"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="8"
                  cy="21"
                  r="1"
                />
                <circle
                  cx="19"
                  cy="21"
                  r="1"
                />
                <path
                  d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex items-center justify-between p-4 lg:px-6"
    >
      <div
        class="flex w-full items-center"
      >
        <div
          class="flex w-full md:w-1/3 items-center"
        >
          <div
            class="mr-2 flex w-full items-center justify-center md:w-auto lg:mr-6"
          >
            <button
              class="inline-flex items-center text-white justify-center text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 underline-offset-4 hover:underline dark:text-neutral-100 h-11 w-11 rounded-md"
            >
              <div
                class="flex flex-none items-center justify-center border border-neutral-700 bg-black h-[40px] w-[40px] rounded-xl"
              >
                <svg
                  aria-label="Acme Store logo"
                  class="h-5 w-5 fill-white h-[16px] w-[16px]"
                  name="logo-icon"
                  viewBox="0 0 32 28"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.5758 9.75769L16 0L0 28H11.6255L21.5758 9.75769Z"
                  />
                  <path
                    d="M26.2381 17.9167L20.7382 28H32L26.2381 17.9167Z"
                  />
                </svg>
              </div>
            </button>
            <a
              class="ml-2 text-sm font-medium uppercase text-white"
              href=""
            >
              acme store
            </a>
          </div>
          <ul
            class="gap-6 text-sm md:flex md:items-center"
          >
            <li
              class="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer"
            >
              All
            </li>
            <li
              class="underline-offset-4 hover:underline text-neutral-400 hover:text-neutral-300 cursor-pointer"
            >
              Shirts
            </li>
          </ul>
        </div>
        <div
          class="justify-center md:flex md:w-1/3"
        >
          <div
            class="relative w-full"
          >
            <input
              class="w-full rounded-lg border border-neutral-800 bg-transparent px-4 py-2 text-sm text-white placeholder:text-neutral-400"
              placeholder="Search Products..."
              type="text"
            />
            <div
              class="absolute inset-y-0 right-0 flex items-center pr-4 pointer-events-none"
            >
              <svg
                aria-hidden="true"
                class="w-4 h-4 text-white"
                fill="none"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="flex justify-end md:w-1/3"
        >
          <button
            class="inline-flex items-center text-white justify-center text-sm font-medium focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 border border-neutral-700 transition-colors h-11 w-11 rounded-md"
          >
            <svg
              class="lucide lucide-shopping-cart"
              fill="none"
              height="16"
              name="cart-icon"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="8"
                cy="21"
                r="1"
              />
              <circle
                cx="19"
                cy="21"
                r="1"
              />
              <path
                d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
